File: boosting/src/learning/regressors/LogisticRegressor.java
package learning.regressors;

import data.ClassifiedData;
import data.Data;
import data.RegressionData;
import utils.LogisticUtils;
import utils.MatrixUtils;
import utils.RandomUtils;

import java.util.List;

public class LogisticRegressor implements Regressor {

    private final int inputSize;

    private final double learningStep;
    private final int iterations;

    private double[] c;

    public LogisticRegressor(int inputSize, double learningStep, int iterations) {
        this.inputSize = inputSize;
        this.learningStep = learningStep;
        this.iterations = iterations;

        this.c = new double[inputSize];
        for (int i = 0; i < inputSize; i++) {
            c[i] = RandomUtils.getInstance().nextDouble() - 0.5;
        }
    }

    @Override
    public void train(List<RegressionData> dataSet) {
        for (int iter = 0; iter < iterations; iter++) {
            double[] sum = new double[inputSize];
            for (RegressionData data : dataSet) {
                double[] x = data.asVector();
                double y = data.output();
                sum = MatrixUtils.add(sum, MatrixUtils.multiply(x, y - regress(data)));
            }
            this.c = MatrixUtils.add(c, MatrixUtils.multiply(sum, learningStep));
        }
    }

    /**
     * Shows probability that class = 1
     */
    @Override
    public double regress(Data data) {
        return f(MatrixUtils.multiply(c, data.asVector()));
    }

    private double f(double z) {
        return LogisticUtils.logisticFunction(z);
    }
}

File: boosting/src/learning/regressors/LogisticRegressor.kt
package learning.regressors

import data.Data
import data.DataWithResult
import learning.model.Classifier
import learning.model.ModelWithTeacher
import utils.LogisticUtils
import utils.MatrixUtils
import utils.RandomUtils

open class LogisticRegressor(private val inputSize: Int, private val learningStep: Double, private val iterations: Int) : ModelWithTeacher {

    private val SELECTION_POINT = 0.5

    private var c: DoubleArray = DoubleArray(inputSize)

    override fun output(data: Data): Double {
        return probability(data)
    }

    override fun train(data: List<DataWithResult>) {
        for (iter in 0..iterations - 1) {
            var sum = DoubleArray(inputSize)
            for (point in data) {
                val x = point.vector
                val y = point.result
                sum = MatrixUtils.add(sum, MatrixUtils.multiply(x, y - probability(point)))
            }
            this.c = MatrixUtils.add(c, MatrixUtils.multiply(sum, learningStep))
        }
    }

    /**
     * Shows probability that class = 1
     */
    private fun probability(data: Data): Double {
        return 2 * f(MatrixUtils.multiply(c, data.vector)) - 1
    }

    private fun f(z: Double): Double {
        return LogisticUtils.logisticFunction(z)
    }
}
